File: y2018/python/day2.py
box_ids = [b.strip() for b in open('input/day2.txt').readlines()]

exactly_two = 0
exactly_three = 0
for b in box_ids:
    occ = {}
    for c in b:
        if c not in occ:
            occ[c] = 0
        occ[c] += 1
    exactly_two += 2 in occ.values()
    exactly_three += 3 in occ.values()

print(f"checksum: {exactly_two * exactly_three}")

found_correct = False
for i in range(len(box_ids)):
    for j in range(i + 1, len(box_ids)):
        diff_num = sum(c1 != c2 for c1, c2 in zip(box_ids[i], box_ids[j]))
        if diff_num == 1:
            common = ''
            for c1, c2 in zip(box_ids[i], box_ids[j]):
                if c1 == c2:
                    common += c1
            print(f"common box id string: {common}")

            found_correct = True

    if found_correct:
        break

File: y2018/rust/src/bin/day2.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let box_ids: Vec<_> = fs::read_to_string("../input/day2.txt")
        .unwrap()
        .split_whitespace()
        .map(|s| s.to_string())
        .collect();

    let char_freq: Vec<HashMap<char, u32>> = box_ids
        .iter()
        .map(|s| {
            s.chars().fold(HashMap::new(), |mut map, val| {
                map.entry(val).and_modify(|frq| *frq += 1).or_insert(1);
                return map;
            })
        })
        .collect();

    let three_letters = char_freq
        .iter()
        .filter(|cf| cf.values().any(|v| *v == 3))
        .count();
    let two_letters = char_freq
        .iter()
        .filter(|cf| cf.values().any(|v| *v == 2))
        .count();
    println!("checksum: {}", three_letters * two_letters);

    'search:
    for s1 in &box_ids {
        for s2 in &box_ids {
            if s1 == s2 {
                continue;
            }
            let diff_num = s1
                .chars()
                .zip(s2.chars())
                .filter(|(c1, c2)| c1 != c2)
                .count();
            if diff_num == 1 {
                let mut res = "".to_string();
                s1.chars()
                    .zip(s2.chars())
                    .filter(|(c1, c2)| c1 == c2)
                    .for_each(|(c, _)| res.push(c));
                println!("common letters: {res}");
                break 'search;
            }
        }
    }
}
